pkg/operator/k8sutil/node.go

```diff
@@ -237,25 +243,27 @@ func GetKubernetesNodesMatchingRookNodes(rookNodes []rookalpha.Node, clientset k
 	return nodes, nil
 }
 
-// RookNodesMatchingKubernetesNodes returns only the given Rook nodes which have a corresponding
-// match in the list of Kubernetes nodes.
-func RookNodesMatchingKubernetesNodes(rookStorage rookalpha.StorageScopeSpec, kubernetesNodes []v1.Node) []rookalpha.Node {
-	nodes := []rookalpha.Node{}
-	for _, kn := range kubernetesNodes {
-		for _, rn := range rookStorage.Nodes {
-			if rookNodeMatchesKubernetesNode(rn, kn) {
-				rn.Name = normalizeHostname(kn)
-
+// RookNodesMatchingKubernetesNodes returns allNodes, matchingNodes
+// allNodes: is a copy of rookStorage.Nodes updated with the following updates:
+//     - node.Name uses hostname label if possible
+//     - node.Location is updated if rookStorage.TopologyAware is true
+// matchingNodes is the subset of allNodes that have a match in the kubeNodes list.
+func RookNodesMatchingKubernetesNodes(rookStorage rookalpha.StorageScopeSpec, kubeNodes []v1.Node) ([]rookalpha.Node, []rookalpha.Node) {
+	matchingNodes := []rookalpha.Node{}
+	for _, kn := range kubeNodes {
+		for rnNum := range rookStorage.Nodes {
+			if rookNodeMatchesKubernetesNode(rookStorage.Nodes[rnNum], kn) {
+				rookStorage.Nodes[rnNum].Name = normalizeHostname(kn)
 				// modify the node Location if topology awareness enabled
 				if rookStorage.TopologyAware == true {
-					rn.Location = nodeTopologyLocation(kn, rn.Location)
+					rookStorage.Nodes[rnNum].Location = nodeTopologyLocation(kn, rookStorage.Nodes[rnNum].Location)
 				}
 
-				nodes = append(nodes, rn)
+				matchingNodes = append(matchingNodes, rookStorage.Nodes[rnNum])
 			}
 		}
 	}
-	return nodes
+	return rookStorage.Nodes, matchingNodes
```

Comment:
Why are you returning `rookStorage.Nodes`? I don't see it used anywhere where this function is called.

Reply:
It's updating the c.DesiredStorage.Nodes of the osd cluster struct.

Reply:
Sure, but why? It wasn't being updated before, what do we gain from updating it now?

Reply:
The location is take from [there](https://github.com/rook/rook/blob/c1e5a46887ffb6b31597b7920a748c2c34302067/pkg/operator/ceph/cluster/osd/osd.go#L551)

Reply:
That doesn't tell me why we're updating it NOW (in this PR) when we weren't updating it before. Does this fix something that was previously broken? From the code you linked, the only thing that's used are the resource requirements on the nodes, which are not being modified or referenced here.

Reply:
It was the main thing that prevented the crush location from propagating.
GetValidNodes used to return the validNodes (now valid desired nodes) with two changes: 

- node.Name value changed to the node's hostname label.
- node.Location populated with failure domain labels.

So it was doing more than just returning a subset of the nodes, it was updating them with info from the k8s objects.

This info was put in c.ValidStorage.Nodes, but it was also needed in c.DesiredStorage.Nodes. It is also not possible to copy them later because the node.Name in c.ValidStorage was not necessarily the name.

Reply:
It seems very wrong that we should be getting any node information from `c.DesiredStorage`, Rook-Ceph should only be using `c.ValidStorage`, just from the name alone. It seems like a bug that `resolveNode()` uses `c.DesiredStorage`. Is it possible to fix that and see if you can get away with only returning/updating `c.ValidStorage`?

Reply:
@travisn I'd also like to get your input here, to see if I'm just getting this all wrong.

Reply:
The DesiredStorage is supposed to represent the static settings in the CR, while the ValidStorage represents what was discovered. In the past the `location` was only assumed to be on the CR, but now we're discovering it dynamically. It really should be combined into the ValidStorage where dynamic info is used, so it does look like we need to change the location handling to use ValidStorage instead of DesiredStorage.

Reply:
I agree with @jarrpa that this doesn't fully make sense. It feels hacky at best.

Reply:
If you look at the comment where `DesiredStorage` is defined, it is this
```go
DesiredStorage  rookalpha.StorageScopeSpec // user-defined storage scope spec
```

DesiredStorage is intended to be a read-only structure containing the user's original requests. This should **never** be modified.